cpp/test/neighbors/ann_cagra.cuh

```diff
@@ -110,42 +112,69 @@ testing::AssertionResult CheckOrder(raft::host_matrix_view<IdxT, int64_t> index_
   return testing::AssertionSuccess();
 }
 
+template <typename T>
+struct fpi_mapper {};
+
+template <>
+struct fpi_mapper<double> {
+  using type                         = int64_t;
+  static constexpr int kBitshiftBase = 50;
+};
+
+template <>
+struct fpi_mapper<float> {
+  using type                         = int32_t;
+  static constexpr int kBitshiftBase = 22;
+};
+
+template <>
+struct fpi_mapper<half> {
+  using type                         = int16_t;
+  static constexpr int kBitshiftBase = 8;
+};
+
 // Generate dataset to ensure no rounding error occurs in the norm computation of any two vectors.
 // When testing the CAGRA index sorting function, rounding errors can affect the norm and alter the
 // order of the index. To ensure the accuracy of the test, we utilize the dataset. The generation
 // method is based on the error-free transformation (EFT) method.
-RAFT_KERNEL GenerateRoundingErrorFreeDataset_kernel(float* const ptr,
+template <typename T>
+RAFT_KERNEL GenerateRoundingErrorFreeDataset_kernel(T* const ptr,
                                                     const uint32_t size,
-                                                    const uint32_t resolution)
+                                                    const typename fpi_mapper<T>::type resolution)
 {
   const auto tid = threadIdx.x + blockIdx.x * blockDim.x;
   if (tid >= size) { return; }
 
-  const float u32 = *reinterpret_cast<const int32_t*>(ptr + tid);
+  const float u32 = *reinterpret_cast<const typename fpi_mapper<T>::type*>(ptr + tid);
   ptr[tid]        = u32 / resolution;
 }
 
+template <typename T>
 void GenerateRoundingErrorFreeDataset(
   const raft::resources& handle,
-  float* const ptr,
+  T* const ptr,
   const uint32_t n_row,
   const uint32_t dim,
   raft::random::RngState& rng,
   const bool diff_flag  // true if compute the norm between two vectors
 )
 {
+  using mapper_type         = fpi_mapper<T>;
+  using int_type            = typename mapper_type::type;
   auto cuda_stream          = resource::get_cuda_stream(handle);
   const uint32_t size       = n_row * dim;
   const uint32_t block_size = 256;
   const uint32_t grid_size  = (size + block_size - 1) / block_size;
 
-  const int32_t resolution =
-    1 << static_cast<unsigned>(std::floor((24 - std::log2(dim) - (diff_flag ? 1 : 0)) / 2));
-  raft::random::uniformInt(
-    handle, rng, reinterpret_cast<int32_t*>(ptr), size, -resolution, resolution - 1);
+  const int_type resolution = std::max<int_type>(
+    int_type{128},
+    int_type{1} << static_cast<unsigned>(
+      std::floor((mapper_type::kBitshiftBase - std::log2(dim) - (diff_flag ? 1 : 0)) / 2)));
+  raft::random::uniformInt<int_type>(
+    handle, rng, reinterpret_cast<int_type*>(ptr), size, -resolution, resolution - 1);
```

Comment:
The transformation is not correct, so the sorting test will fail depending on the random numbers generated. However, this method of generating random numbers, which results in rounding errors, cannot be used for FP16 when the dimension of the vector is greater than 256, because there is no range of random numbers that allows us not to introduce rounding errors.

I have three possible ways to solve this problem.
1. Skip the sorting test for FP16 when the dimension is greater than 256.
2. Use std::max like the PR code. This does not guarantee that there will be no rounding error, so the test will fail depending on the random numbers generated.
3. Stop using the no rounding error dataset and compute the norm of a vector in the test code with the same order of addition as in the CAGRA implementation. I didn't use this method in the current test implementation because I thought the test and the CAGRA implementation should be independent.

Reply:
Thanks for checking this! I think, the (1) is the most reasonable solution; I'll just add the GTEST_SKIP when the resolution is 1 bit or less.